### PR TITLE
fix error when logging to progress bar with reserved name

### LINKED ISSUE #5299: self.log("loss", loss, prog_bar=True) leads to an error
## 🐛 Bug

There is a confusing inconsistency with the special-casing of `loss`: you normally don't need to log it for it to show up in the progress bar, but you need to log it to use it as a `ModelCheckpoint` `monitor=` key. But if you log it with `prog_bar=True`, an error is thrown.

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/15YTx4akWcl10Zo0VMW6jPSYa4wxZeXac?usp=sharing

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

### Expected behavior

`loss` should either behave like other metrics (ie. require `self.log` to show up in the progress bar), or its special casing should be consistent (ie. it should be available as a monitor key for ModelCheckpoint automatically). Additionally, "double logging" to progress bar should not cause an error - it should be de-duplicated.

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cu101
	- pytorch-lightning: 0.10.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #5299

The statement

```python
self.log("loss", loss, prog_bar=True)
```

currently results in an error, because the loss is already logged to the progress bar by default. 
Instead of an error message, we overwrite the progress bar dict entry by what the user logged and print a warning.

The warning was suggested by @edenlightning. Alternatively, we could also just overwrite the value without the warning because the user explicitly calls `self.log` anyway.

### CODE REVIEW

File: pytorch_lightning/trainer/properties.py [hunk 1]
```diff
@@ -185,7 +185,21 @@ def progress_bar_dict(self) -> dict:
...
+        standard_metrics = ref_model.get_progress_bar_dict()
+        logged_metrics = self.logger_connector.progress_bar_metrics
+        duplicates = list(standard_metrics.keys() & logged_metrics.keys())
+        if duplicates:
+            rank_zero_warn(
+                f"The progress bar already tracks a metric with the name '{duplicates[0]}' and"
+                f" `self.log('{duplicates[0]}', ..., prog_bar=True)` will overwrite this value. "
+                f" If this is undesired, change the name or override `get_progress_bar_dict()`."
+                f" in LightingModule.",
+                UserWarning
```
Comment (awaelchli) [inline pytorch_lightning/trainer/properties.py:198]: alternatively, one could drop this warning and just overwrite the default metric with what the user logged, since it is deliberate.

File: tests/trainer/logging/test_logger_connector.py [hunk 2]
```diff
@@ -425,3 +425,21 @@ def test_dataloader(self):
     )
     trainer.fit(model)
     trainer.test(model, ckpt_path=None)
+
+
+def test_logging_to_progress_bar_with_reserved_key(tmpdir):
```
Comment (awaelchli) [inline tests/trainer/logging/test_logger_connector.py:430]: @tchaton @Borda Where should I put this test? I'm not familiar with how the logging tests are structured.

File: pytorch_lightning/trainer/properties.py [hunk 3]
```diff
@@ -185,7 +185,21 @@ def progress_bar_dict(self) -> dict:
         """ Read-only for progress bar metrics. """
         ref_model = self.model if not self.data_parallel else self.model.module
         ref_model = cast(LightningModule, ref_model)
-        return dict(**ref_model.get_progress_bar_dict(), **self.logger_connector.progress_bar_metrics)
+
+        standard_metrics = ref_model.get_progress_bar_dict()
+        logged_metrics = self.logger_connector.progress_bar_metrics
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/properties.py:190]:
```suggestion
        logged_metrics = self.progress_bar_metrics
```

File: pytorch_lightning/trainer/properties.py [hunk 4]
```diff
@@ -185,7 +185,21 @@ def progress_bar_dict(self) -> dict:
         """ Read-only for progress bar metrics. """
         ref_model = self.model if not self.data_parallel else self.model.module
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/properties.py:186]:
```suggestion
        ref_model = self.get_model()
```
so for ddp ref_model is just model and not model.module???

File: pytorch_lightning/trainer/properties.py [hunk 5]
```diff
@@ -185,7 +185,21 @@ def progress_bar_dict(self) -> dict:
...
         ref_model = cast(LightningModule, ref_model)
-        return dict(**ref_model.get_progress_bar_dict(), **self.logger_connector.progress_bar_metrics)
+
+        standard_metrics = ref_model.get_progress_bar_dict()
+        logged_metrics = self.logger_connector.progress_bar_metrics
+        duplicates = list(standard_metrics.keys() & logged_metrics.keys())
+        if duplicates:
+            rank_zero_warn(
+                f"The progress bar already tracks a metric with the name '{duplicates[0]}' and"
+                f" `self.log('{duplicates[0]}', ..., prog_bar=True)` will overwrite this value. "
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/properties.py:195]: shall it have all the duplicate keys? for eg `v_num` and `split_idx` are also there

File: pytorch_lightning/trainer/properties.py [hunk 6]
```diff
@@ -185,7 +185,21 @@ def progress_bar_dict(self) -> dict:
...
+                f" If this is undesired, change the name or override `get_progress_bar_dict()`."
+                f" in LightingModule.",
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/properties.py:197]:
```suggestion
                f" If this is undesired, change the name or override `get_progress_bar_dict()`"
                f" in `LightingModule`.",
```

**Review by rohitgr7 — COMMENTED** [review]
IMO at least adding a warning is better rather than silently updating it.

File: pytorch_lightning/trainer/properties.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:186]: no but there is some shenanigans going on in the trainer that will make it so that self.model is the unwrapped model. 
This whole model retrieval will be addressed and made more transparent by our refactoring in #4510

File: pytorch_lightning/trainer/properties.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:195]: it has nothing to do with v_num and split_idx. 
Here I'm just simplifying the message to one occurence. For 2 duplicates, the user would see the first one and then after fixing it the second time the next message

File: pytorch_lightning/trainer/properties.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:198]: @rohitgr7 if the user wants their logged value in the progress bar, there would be no way to turn off the warning. That's the main argument against the warning, because the user is deliberately choosing to display that value with `prog_bar=True`.

File: pytorch_lightning/trainer/properties.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/properties.py:195]: not even doing it this way `f"with names {', '.join(duplicates)} and self.log('duplicates[0]', ...)"`, just to avoid the possibility of second time? one can override `get_progress_bar_dict` hook and add a key there and also log it somewhere else too. Although this case is super rare.

File: tests/trainer/logging/test_logger_connector.py (hunk 2 above)
Comment (Borda) [reply tests/trainer/logging/test_logger_connector.py:430]: tests/trainer/logging_process ?

Comment (awaelchli) [reply tests/trainer/logging/test_logger_connector.py:430]: this folder does not exist on master, only dev branch

File: pytorch_lightning/trainer/properties.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/properties.py:195]: ok I did it

File: tests/trainer/logging/test_logger_connector.py (hunk 2 above)
Comment (Borda) [reply tests/trainer/logging/test_logger_connector.py:430]: oh, I see.... :]

**Review by Borda — APPROVED** [review]
nice one :]

File: tests/trainer/logging/test_logger_connector.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging/test_logger_connector.py:430]: Maybe callback/progress ?

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/trainer/properties.py [hunk 7]
```diff
@@ -183,9 +183,23 @@ def progress_bar_callback(self):
...
+        ref_model = self.get_model()
         ref_model = cast(LightningModule, ref_model)
-        return dict(**ref_model.get_progress_bar_dict(), **self.logger_connector.progress_bar_metrics)
+
+        standard_metrics = ref_model.get_progress_bar_dict()
+        logged_metrics = self.progress_bar_metrics
+        duplicates = list(standard_metrics.keys() & logged_metrics.keys())
+        if duplicates:
+            rank_zero_warn(
+                f"The progress bar already tracks a metric with the name '{', '.join(duplicates)}' and"
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/properties.py:194]:
```suggestion
                f"The progress bar already tracks a metric with the name(s) '{', '.join(duplicates)}' and"
```

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/logging/test_logger_connector.py [hunk 8]
```diff
@@ -425,3 +425,21 @@ def test_dataloader(self):
...
+            output = super().training_step(*args, **kwargs)
+            self.log("loss", output["loss"], prog_bar=True)
+            return output
+
+    model = TestModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_steps=2,
+    )
+    with pytest.warns(UserWarning, match="The progress bar already tracks a metric with the name 'loss'"):
```
Comment (rohitgr7) [inline tests/trainer/logging/test_logger_connector.py:444]:
```suggestion
    with pytest.warns(UserWarning, match="The progress bar already tracks a metric with the name(s) 'loss'"):
```

Comment (awaelchli) [reply tests/trainer/logging/test_logger_connector.py]: the parentheses will not work, this is a regular expression
needs to be a raw string I think

**Comment by carmocca** [comment]
@awaelchli can you move this test to `logging_tests` directory?. For the reasons described in https://github.com/PyTorchLightning/pytorch-lightning/pull/4451 (and https://github.com/PyTorchLightning/pytorch-lightning/pull/5452#pullrequestreview-565300740)

We should be more mindful about where we place our tests

**Comment by awaelchli** [comment]
> We should be more mindful about where we place our tests

I asked but was not fast enough to react before it was merged. I am mindful and I have suggested several times before to define a consistent testing file/folder structure. Here #5667
